aiida/tools/importexport/dbexport/__init__.py

```diff
@@ -633,25 +624,43 @@ def _write_group_mappings(*, group_pks: Set[int], batch_size: int, writer: Archi
 
 
 def _write_node_repositories(
-    *, node_pks: Set[int], node_pk_2_uuid_mapping: Dict[int, str], writer: ArchiveWriterAbstract
+    *, node_pks: Set[int], repository_metadata_mapping: Dict[int, dict], writer: ArchiveWriterAbstract
 ):
     """Write all exported node repositories to the archive file."""
     with get_progress_reporter()(total=len(node_pks), desc='Exporting node repositories: ') as progress:
 
-        for pk in node_pks:
+        with tempfile.TemporaryDirectory() as temp:
```

Comment:
control should be inverted here; you should just pass `container_profile` and `hashkeys` to the writer then it can decide how to extract the data and store it

Reply:
Not sure exactly what you want me to do here. Do you mean I have to change `writer.write_repository_container` from `writer.write_repository_container(container_export)` to `writer.write_repository_container(container_profile, hashkeys)`? This won't change that much though, it will simply move the line `container_profile.export(set(hashkeys), container_export, compress=True)` into the `writer.write_repository_container` method.

Now I can see that if you want to implement the copying of the file extraction manually, but I don't think we should do this since the most efficient way is what is implemented by `container.export` provided by the `disk-objectstore` itself. Of course we could make a manual implementation, but this maybe very inefficient

Reply:
> This won't change that much though, it will simply move the line container_profile.export(set(hashkeys), container_export, compress=True) into the writer.write_repository_container method.

no it won't change anything, but the point is the control of how to write the repository to the archive file should be with the writer, not the export function.
this allows for anyone to create a bespoke writer that writes it however they want, without it being effectively hard-coded in the `export` function.

(Just in case you hadn't heard of it before, this is what I'm referring to when I say inverting control: https://en.wikipedia.org/wiki/Inversion_of_control, and was very much a goal with all my refactoring of the import/export code)